### PR TITLE
Created events to dispatch in JWTGuard

### PR DESCRIPTION
**Pull request by GabrielDeveloper**
PR to #886

### CODE REVIEW

**Comment by GabrielDeveloper** [comment]
@tymondesigns Could you see this PR please? Thank you so much!

File: src/JWTGuard.php [hunk 1]
```diff
@@ -121,6 +135,10 @@ public function attempt(array $credentials = [], bool $login = true)
         $this->lastAttempted = $user = $this->provider->retrieveByCredentials($credentials);
 
         if ($this->hasValidCredentials($user, $credentials)) {
+            $this->events->dispatch(
+                new JWTAttempt($user)
+            );
```
Comment (tymondesigns) [inline src/JWTGuard.php]: I think this should be dispatched regardless of valid credentials

File: src/Events/JWTInvalidate.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+     * @param  \Illuminate\Contracts\Auth\Authenticatable  $user
+     * @param  \Tymon\JWTAuth\Token|string  $token
+     *
+     * @return void
+     */
+    public function __construct($user, $token)
+    {
+        $this->token = $token;
+        $this->user = $user;
+    }
```
Comment (tymondesigns) [inline src/Events/JWTInvalidate.php]: All of these events should extend a common abstract class containing the constructor and properties

**Review by tymondesigns — CHANGES_REQUESTED** [review]
Thanks for this, see my comments

File: src/JWTGuard.php (hunk 1 above)
Comment (GabrielDeveloper) [reply src/JWTGuard.php]: I agree, ready!

File: src/Events/JWTInvalidate.php (hunk 2 above)
Comment (GabrielDeveloper) [reply src/Events/JWTInvalidate.php]: Ready, thank you!

**Comment by GabrielDeveloper** [comment]
@tymondesigns all requested changes are ready, thank one more time!

**Comment by dhcmega** [comment]
Hi, will this then trigger the Login event? How can I install this patch?
In my composer json I have: `"tymon/jwt-auth": "^1.0"`
thanks!

**Comment by GabrielDeveloper** [comment]
@dhcmega, yes.
This PR will implement the Login Event, but has not yet been merged.

When this PR is merged, you should use the 2.0.x-dev tag while 2.0 is not ready.

@tymondesigns Could you see the yours request changes, please? Thanks.

**Comment by dhcmega** [comment]
Hi @GabrielDeveloper thanks for your reply.
Is 2.0-dev production ready? If that is the case, I will upgrade now to 2.
(it might sound obvious, but there are plenty of dev packages being used for prod)
Thanks!

**Comment by tymondesigns** [comment]
Please do not use 2.0 in production

**Comment by GabrielDeveloper** [comment]
Up =)

**Comment by tymondesigns** [comment]
Thanks 👍

**Comment by cubitworx** [comment]
Hi @tymondesigns Is there a roadmap for this package? Any idea when the expected release date for 2.x will be? I need access to these events. Alternatively, any chance this set of changes can be applied to 1.x? I don't mind doing a PR. Thanks in advance

Update: In hindsight, I have just gone through this implementation & it doesn't cover all of the authentication events (and I am not even sure why JWT specific events were created as opposed to using the Laravel defaults) so I'll have to go with a custom abstraction. Thanks anyway

**Comment by lucianobargmann** [comment]
I cannot agree more with you @cubitworx - WHY arent we just throwing Laravel events? What is the reason for creating custom events?

Seriously, I spent a freaking week trying to figure out why my events were not being fired. And yes, let's back port this to 1.x. 

None of the other options are feasible, once I do not have a UI on my project. I will probably fork this and change the 5 lines needed.

**Comment by francoism90** [comment]
@GabrielDeveloper Is it possible to merge this in `1.0.0-rc.*` as well and simple add the auth events?

**Comment by GabrielDeveloper** [comment]
@francoism90, @cubitworx I created a PR to add these events on 1.0.* version: #1887 

Let's wait @tymondesigns to accept! Thanks!

**Comment by GabrielDeveloper** [comment]
@lucianobargmann the events were created because we can have 2 ways to authentication on our system.

For example, we can have the default authentication to log in the system normally and we can have too an API with JWT auth to interact with mobile apps.

So I think that is a good idea we have our own events to manipulate the way that we want

**Comment by lucianobargmann** [comment]
To be honest with you, I forked and solved my problem. Thanks for all the
work! Let me know if I can be of service.

Em qua, 9 de out de 2019 18:18, Gabriel Gonçalves Rodrigues <
notifications@github.com> escreveu:

**Comment by nguyenphuocnhatthanh** [comment]
Tks for PR

**Comment by cubitworx** [comment]
Thanks for the explanation @GabrielDeveloper but I would have at least expected the package to have Laravel events triggered in addition to your events (if need be); since you are targeting Laravel & it is an auth replacement for which Laravel has implemented a standard to work from (albeit not to your liking). The events that you have created don't really make sense to me (maybe lack of documentation) or cover all the events I would need / expect.

I ended up overloading the methods of the package I needed to in order to trigger the events.

PS. I haven't checked the PR & don't know if anything has changed from the implementation I was originally commenting on.

**Comment by keyboardSmasher** [comment]
There's no need for custom events, you can just call
```
event(new Login('jwt', $user, false));
```
from `JWTGuard::login` for instance.

Then in your listener you'd have
```
$event->guard === 'jwt'
```

**Comment by cubitworx** [comment]
@keyboardSmasher Having to trigger the events yourself is tantamount to creating custom events... you're trying to solve the problem of the library not fire the default auth events that anyone using Laravel would expect to listen for.

The point of using Laravel defined events is so that you end up with less vendor lock-in, as is the case when needing to use custom events for auth.

In most instances you don't care what guard triggered the event, you simply want to action something when that event happens, so this code is arguably hardly ever needed because it adds to vendor lock-in:
```
$event->guard === 'jwt'
```

**Comment by keyboardSmasher** [comment]
@cubitworx
What I did was create a new class:
```
use Tymon\JWTAuth\JWTGuard;
use Illuminate\Auth\Events\Login;
use Tymon\JWTAuth\Contracts\JWTSubject;

class ExtendedJWTGuard extends JWTGuard
{
    public function login(JWTSubject $user)
    {
        $token = $this->jwt->fromUser($user);
        $this->setToken($token)->setUser($user);
        event(new Login('api', $user, false));

        return $token;
    }
}
```
Then in `AuthServiceProvider`:
```
    public function boot()
    {
        $this->app['auth']->extend('jwt', function ($app, $name, array $config) {
            $guard = new ExtendedJWTGuard(
                $app['tymon.jwt'],
                $app['auth']->createUserProvider($config['provider']),
                $app['request']
            );

            $app->refresh('request', $guard, 'setRequest');

            return $guard;
        });
```
Now I can listen for Laravel's `Login` event and differentiate between `web` and `api` in my event listener.
```
$event->guard === 'api'
```
I don't know...it works for me.

**Comment by caryxiao** [comment]
> @cubitworx
> What I did was create a new class:
> 
> ```
> use Tymon\JWTAuth\JWTGuard;
> use Illuminate\Auth\Events\Login;
> use Tymon\JWTAuth\Contracts\JWTSubject;
> 
> class ExtendedJWTGuard extends JWTGuard
> {
>     public function login(JWTSubject $user)
>     {
>         $token = $this->jwt->fromUser($user);
>         $this->setToken($token)->setUser($user);
>         event(new Login('api', $user, false));
> 
>         return $token;
>     }
> }
> ```
> 
> Then in `AuthServiceProvider`:
> 
> ```
>     public function boot()
>     {
>         $this->app['auth']->extend('jwt', function ($app, $name, array $config) {
>             $guard = new ExtendedJWTGuard(
>                 $app['tymon.jwt'],
>                 $app['auth']->createUserProvider($config['provider']),
>                 $app['request']
>             );
> 
>             $app->refresh('request', $guard, 'setRequest');
> 
>             return $guard;
>         });
> ```
> 
> Now I can listen for Laravel's `Login` event and differentiate between `web` and `api` in my event listener.
> 
> ```
> $event->guard === 'api'
> ```
> 
> I don't know...it works for me.

I think it works. 😀

**Pull request merged**